File: Data Structures & Algorithms/validate-parentheses/submission-2.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(char c : s)
        {
            if(c == '(' || c==')'|| c=='[' || c==']'|| c=='{' || c=='}')
            {

                    if(c==')' || c==']'|| c=='}')
                    {
                        if(!st.empty())
                        {
                            if(c==')')
                            {
                                if(st.top() != '(')
                                    return false;
                            }
                            else if(c==']')
                            {
                                if(st.top() != '[')
                                {
                                    return false;
                                }
                            }
                            else if(c=='}')
                            {
                                cout<< st.top() << endl;
                                if(st.top() != '{'){
                                    return false;
                                }
                            }
                            st.pop();
                        }
                        else
                            return false;

                    }
                    else
                    {
                        st.push(c);
                    }

            }
        }
        if(!st.empty())
            return false;
        return true;
        
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        maxArea = 0;
        for i,val in enumerate(heights):
            start =i; 
            while stack and stack[-1][1] > val:
                index, height = stack.pop()
                width = (i - index)
                area = height*width
                if area >= maxArea:
                    maxArea = area
                start = index
            stack.append((start,val))
        for i, h in stack:
            maxArea = max(maxArea, h*(len(heights)-i))
        return maxArea

                

                

                


        